src/filters/ES6Ctors.ml

```diff
@@ -0,0 +1,225 @@
+(*
+	The Haxe Compiler
+	Copyright (C) 2005-2019  Haxe Foundation
+
+	This program is free software; you can redistribute it and/or
+	modify it under the terms of the GNU General Public License
+	as published by the Free Software Foundation; either version 2
+	of the License, or (at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *)
+open Common
+open Globals
+open Type
+open Texpr.Builder
+
+(* name of the method to which the constructor is extracted *)
+let ctor_method_name = "_hx_constructor"
+
+(* name of the static bool flag to skip constructor body execution *)
+let skip_ctor_flag_name = "_hx_skip_constructor"
+
+(* replace super(a,b,c) with super._hx_constructor(a,b,c) *)
+let rec replace_super_call e =
+	match e.eexpr with
+	| TCall ({ eexpr = TConst TSuper } as e_super, args) ->
+		let e_super_hxctor = { e_super with eexpr = TField (e_super, FDynamic ctor_method_name) } in
+		{ e with eexpr = TCall (e_super_hxctor, args) }
+	| _ ->
+		map_expr replace_super_call e
+
+exception Accessed_this of texpr
+
+(* return whether given expression has `this` access before calling `super` *)
+let has_this_before_super e = 
+	let rec loop e =
+		match e.eexpr with 
+		| TCall ({ eexpr = TConst TSuper }, args) ->
+			List.iter loop args;
```

Comment:
Hm, maybe, I'll look into it. I got that habit to write exception-driven control flow in OCaml from @Simn :)

Reply:
You mean something like this?

```ocaml
let has_this_before_super e = 
	let this_access = ref None in
	let rec loop e =
		match e.eexpr with 
		| TCall ({ eexpr = TConst TSuper }, args) ->
			List.exists loop args;
		| TConst TThis ->
			this_access := Some e;
			true
		| _ ->
			Type.check_expr loop e
	in
	ignore (loop e);
	!this_access
```

Reply:
Throwing exceptions is pretty good for these kind of tree explorations because you want to exit eagerly. Also, exceptions are fairly cheap in OCaml, plus it's quite idiomatic.

Reply:
But it's not type safe )
But ok.

Reply:
In this case it'll also exit eagerly, right?

Reply:
`Type.check_expr` doesn't seem to guarantee the execution order for the callback, so it won't work like that, so I'm gonna leave this exception-driven.

Reply:
> `Type.check_expr` doesn't seem to guarantee the execution order for the callback

It should given that it's just a bunch of `||` checks.

Reply:
Well there was some issue and I didn't dig much deeper, assuming there's something wrong with predicate execution order, because the original version doesn't look bad too IMO.